src/main/java/net/fabricmc/loom/task/service/TinyRemapperService.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2021 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.task.service;
+
+import java.io.File;
+
+import org.gradle.api.Project;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+
+import net.fabricmc.loom.LoomGradleExtension;
+import net.fabricmc.tinyremapper.TinyRemapper;
+import net.fabricmc.tinyremapper.extension.mixin.MixinExtension;
+
+public abstract class TinyRemapperService implements BuildService<TinyRemapperService.Params>, AutoCloseable {
+	public interface Params extends BuildServiceParameters {
+		ListProperty<Provider<MappingsService>> getMappings();
+
+		// classpath passed into the TinyRemapper instance
+		ConfigurableFileCollection getClasspath();
+
+		Property<Boolean> getUseMixinExtension();
+	}
+
+	public static synchronized Provider<TinyRemapperService> create(Project project, String from, String to) {
+		final LoomGradleExtension extension = LoomGradleExtension.get(project);
+
+		return project.getGradle().getSharedServices().registerIfAbsent(extension.getMappingsProvider().getBuildServiceName("remapJarService", from, to), TinyRemapperService.class, spec -> {
+			spec.parameters(params -> {
+				params.getClasspath().plus(project.getConfigurations().getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME));
```

Comment:
Shouldn't the classpath rather come from `remapJar`'s configuration and not be hardcoded? Also, this makes it hard to declare some custom `RemapJarTask`s since the classpath is locked to always have these files.